### PR TITLE
Add heFFTe

### PR DESCRIPTION
**Pull request by ax3l**
Add the heFFTe library, a Highly Efficient FFT for Exascale.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] ~If static libraries are linked in, the license of the static library is packaged.~
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by ax3l** [comment]
@conda-forge/help-c-cpp this new package for heFFTe is ready for review.

In the Linux test, I do not understand the linker missing the `libpthread_nonshared.a` library. This might be an MPI or FFTW package bug in CF... I cannot say for sure.
```
[ 20%] Linking CXX executable heffte_example_fftw
/home/conda/staged-recipes/build_artifacts/heffte_1718162756113/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/bin/../lib/gcc/x86_64-conda-linux-gnu/12.3.0/../../../../x86_64-conda-linux-gnu/bin/ld: cannot find /usr/lib64/libpthread_nonshared.a: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [examples/CMakeFiles/heffte_example_fftw.dir/build.make:107: examples/heffte_example_fftw] Error 1
```

File: recipes/heffte/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    Security Administration (NNSA). HeFFTe delivers algorithms for distributed
+    fast-Fourier transforms in on a heterogeneous systems, targeting the
+    upcoming exascale machines.
+
+  doc_url: https://icl-utk-edu.github.io/heffte/
+  dev_url: https://github.com/icl-utk-edu/heffte
+
+extra:
+  recipe-maintainers:
+    - ax3l
```
Comment (ax3l) [inline recipes/heffte/meta.yaml:75]: https://github.com/icl-utk-edu/heffte/issues/48#issuecomment-2162725230
```suggestion
    - ax3l
    - mkstoyanov
```

**Comment by ax3l** [comment]
ping @conda-forge/help-c-cpp any idea on this pthread issue in tests? :)

**Comment by mkstoyanov** [comment]
I agree to be listed as feedstock maintainer.

**Comment by ax3l** [comment]
ping @conda-forge/help-c-cpp any idea on this pthread issue in tests? :)

File: recipes/heffte/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    - {{ mpi }}
+
+test:
+  requires:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ stdlib("c") }}
+    - cmake
+    - make  # [unix]
+    - pkgconfig
```
Comment (ax3l) [inline recipes/heffte/meta.yaml:53]: Might help with the failing test on Linux?
```suggestion
    - pkgconfig
   -  pthread-stubs  # [linux]
```

File: recipes/heffte/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+   -  pthread-stubs  # [linux]
```
Comment (ax3l) [inline recipes/heffte/meta.yaml:54]:
```suggestion
    -  pthread-stubs  # [linux]
```

File: recipes/heffte/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+test:
+  requires:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ stdlib("c") }}
+    - cmake
+    - make  # [unix]
+    - pkgconfig
+    -  pthread-stubs  # [linux]
```
Comment (ax3l) [inline recipes/heffte/meta.yaml:54]:
```suggestion
    - pkgconfig
    - pthread-stubs  # [linux]
  host:
    - fftw * mpi_{{ mpi }}_*
    - {{ mpi }}
  run:
    - fftw * mpi_{{ mpi }}_*
    - {{ mpi }}

test:
  requires:
    - {{ compiler("c") }}
    - {{ compiler("cxx") }}
    - {{ stdlib("c") }}
    - cmake
    - make  # [unix]
    - pkgconfig
```

**Comment by ax3l** [comment]
@conda-forge/fftw do you know what might be the issue here? I am trying to use FFTW to build a dependency and on Linux it fails with missing static pthread dependencies.

**Comment by egpbos** [comment]
Hey Axel, sorry, no idea, did you figure it out in the meantime?

**Comment by jakirkham** [comment]
Maybe [this SO question]( https://stackoverflow.com/q/71340058 ) is helpful?

Also in this issue ( https://github.com/conda-forge/stir-feedstock/issues/101 ) a similar problem was encountered

**Comment by ax3l** [comment]
Thank you for your help. Argh, this one is tricky. It's not an isolated build with pip, just a regular CMake build.

I wonder if I should manually remove the "accidentally" linked pthreads in the build scripts from the image to make sure they are not picked up as a work-around...

I am not sure if this comes in propagated from a dependency. I cannot spot heFFTe [searching for pthreads](https://github.com/search?q=repo%3Aicl-utk-edu%2Fheffte+threads&type=code) on its own, but FFTW or MKL from other packages might pull this in...

**Comment by ax3l** [comment]
Ok, with  GNU compilers there is an unconditional search for OpenMP in the FFTW heFFTe logic, let me add those for linux to the dependencies:
https://github.com/icl-utk-edu/heffte/blob/v2.4.0/cmake/FindHeffteFFTW.cmake#L78-L80

File: recipes/heffte/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ stdlib("c") }}
+    - cmake
+    - make  # [unix]
+    - pkgconfig
+    - pthread-stubs  # [linux]
```
Comment (ax3l) [inline recipes/heffte/meta.yaml:39]:
```suggestion
    - libgomp  # [linux]
```

**Comment by ax3l** [comment]
I think the FFTW library is overlinking in the `FindHEFFTW.cmake` the FFTW library variants... this looks too much to me - threads and OMP versions at the same time?
```
../libheffte.so.2.4.0 $PREFIX/lib/libmpi.so $PREFIX/lib/libfftw3.so $PREFIX/lib/libfftw3f.so $PREFIX/lib/libfftw3_omp.so $PREFIX/lib/libfftw3f_omp.so $PREFIX/lib/libfftw3_threads.so $PREFIX/lib/libfftw3f_threads.so $PREFIX/lib/libgomp.so -lpthread
```

**Comment by ax3l** [comment]
That is possible, I myself do some quite elaborate back-and-forth when searching for FFTW downstream in other codes: https://github.com/ECP-WarpX/WarpX/blob/development/cmake/dependencies/FFT.cmake

@mkstoyanov do you have any recommendations on how to build heFFTe against FFTW on Linux or have seen this issue before?

**Comment by mkstoyanov** [comment]
@ax3l I don't know why I missed the notification on this mention. Feel free to email me at work and I seem to be getting the comments on the heFFTe issues.

I've only ever build heFFTe with FFTW on Linux using either default installed packages from `apt` or the system modules on the HPC clusters. The modules lead only one set of libraries at a time (usually) and I've had no conflicts with apt.

The long list of "overlinking" was a work-around when I don't know what will be installed. The hope is that the compiler will just pick the first one it finds and ignores the rest. I don't know what is the right logic to handle this if we have multiple libraries (e.g., openmp and threads) and how to detect whether a compiler supports OpenMP (clang can be iffy with OpenMP).

The last-resort solution that heFFTe supports is that the user can manually specify:
```
-D FFTW_LIBRARIES=...
-D FFTW_INCLUDES=...
```
If those two are present, heFFTe will not do any "guessing" or try to find or enable anything else.

I don't know how/if you can get the list of libraries and includes from conda but if you can, this should just work.

**Comment by ax3l** [comment]
Thanks, @mkstoyanov!

I am trying to find out: is FFTW thread support (which is optional) are requirement for heFFTe to build? I see a few  calls in a test like this:
https://github.com/icl-utk-edu/heffte/blob/a6236b06f321e06c23f2615a26da1ee82c63b962/benchmarks/speed3d.h#L49-L53
that assumes FFTW presence is equal to thread or OMP support.

**Comment by ax3l** [comment]
@egpbos @jakirkham we made it - heFFTe ready for review & merge :)

**Comment by egpbos** [comment]
Awesome @ax3l! I'm not a reviewer here, but it LGTM nonetheless ;)

**Comment by ax3l** [comment]
Oops, sorry 😅

@conda-forge/help-c-cpp heFFTe package is ready to gooo 🚀

File: recipes/heffte/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  string: {{ mpi_prefix }}_h{{ PKG_HASH }}_{{ build }}  # [not win]
+
+requirements:
+  build:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ stdlib("c") }}
+    - cmake
+    - make  # [unix]
+    - pkgconfig
```
Comment (isuruf) [inline recipes/heffte/meta.yaml:42]:
```suggestion
    - pkg-config
```

File: recipes/heffte/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    - {{ mpi }}               # [not win]
+
+test:
+  requires:
+    - {{ compiler("c") }}
+    - {{ compiler("cxx") }}
+    - {{ stdlib("c") }}
+    - cmake
+    - make  # [unix]
+    - pkgconfig
```
Comment (ax3l) [inline recipes/heffte/meta.yaml:58]:
```suggestion
    - pkg-config
```

File: recipes/heffte/meta.yaml (hunk 6 above)
Comment (ax3l) [reply recipes/heffte/meta.yaml]: Thanks, goo catch

**Comment by ax3l** [comment]
Thank you, Isuru!

**Pull request merged**